src/ne.1.in
.\" @configure_input@
.\"
.\"    Copyright (C) 1999,2000 SASKI Shunsuke.
.\"    All rights reserved. 
.\"
.TH NE 1 "November 15, 1999"
.de sh
.br
.PP
\fB\\$1\fR
.PP
..
.SH 名称
ne \- テキストエディタ
.\"
.\"
.SH	書式
\fBne\fP [\fB-ejor\fP] [\fB-D \fIname\fP=\fIval\fP]
[\fIfile ...\fP] [\fI+line\fP]
.\"
.\"
.SH	解説
.B ne
はモードレスなフルスクリーンのテキストエディタです。
主要な操作や編集方式を MS-DOS系OS で有名な
.B Vz
互換にしていますので、MS-DOS系OS や Windows などで
.B Vz\fP、
.B Mifes
など
.B wordstar
系テキストエディタを使っていた方には使いやすいと思います。

オプションは以下の通り:
.IP \fB\-j\fP:
メッセージを日本語にします。
.IP \fB\-e\fP:
メッセージを英語にします。但し不完全です。
.IP \fB\-o\fp:
オープニングを表示します。
.IP \fB\-r\fP:
プロファイルが保存されていればその時の編集状態に戻します。
.IP \fB-D\fIname\fP=\fIval\fP:
内部変数
.I name
に
.I val
を定義します。
.IP \fB+\fP\fIline\fP:
編集開始する時の行番号を
.B vi
互換方式で指定します。
.\" [-FRSw]
.\"   -F: backupファイルを作成しません。
.\"   -R: readonly (vim: -v)
.\"   -S: secure mode
.\"   -w: 行数指定
.\"   -b: binary
.br
.PP
ファイルを指定しない場合はカレントディレクトリの
.B untitled.txt
というファイルを編集します。
.\"
.\"
.SH 画面構成
.B ne
の画面は最上行以外は基本的にエディタ画面です。
最上行は編集ファイルに対する情報を表示するシステムガイド行です。
システムガイド行を左から説明します。
.\"
.sh ページングモード
左端の一文字はページ切替えのキー操作でどういう挙動をするかを示しています。

.nf
\fBP\fP:	通常のページングモード
\fBS\fP:	文字列検索モード
.fi
.sh インサートモード
次の一文字は文字を打ち込んだ時に上書きするのか、
挿入するのかを示しています。

.nf
\fBi\fP:	挿入モード
\fBR\fP:	上書きモード
.fi
.sh 現在編集行
次の "\fB00001/  1%\fP" のような表示は現在編集中の行を示します。
.B /
の前にある数字は、現在カーソルのある行を \fB1\fPから数えた行番号で、
後ろの数字はその行がファイル全体の大体何％位の位置かを示します。
.sh 現在編集位置。
次の "\fB001[0020]\fP" のような表示は現在編集中の位置 (桁) を示します。
最初の数字がカーソルにある桁位置を \fB0\fPから数えた桁番号で示します。
[]の中の数字はカーソル位置の文字の文字コードを示し、
2ストローク入力の時はその一文字目のコードを示します。
.sh 編集中のファイルのサイズ。
次の数字は編集中のファイルサイズを示しています。
この数値は実際にバッファにある文字のバイト数なのですが、
EUC/LF の場合で計算しているので、
JIS/SJIS モードの場合や
CRLF モードの場合にサイズは食い違ってしまいます。
.sh 現在の対象ファイル情報
最後の "\fB1EL:/home/README.jp\fP" のような表示は編集対象の情報を示します。
数字は編集しているファイルの番号で次の二文字が
編集中のファイルの文字コードと改行コードの種類を示します。

.nf
\fBE\fP:	EUC
\fBS\fP:	SJIS
\fBJ\fP:	JIS

\fBL\fP:	LF
\fB+\fP:	CRLF
\fBC\fP:	CR

その次の一文字はそのファイルを既に変更したかどうかを示します。

\fB:\fP:	まだ変更していない。
\fB*\fP:	既に変更している。

そして最後にフルパスで示したファイルネームが続きます。
.\"
.\"
.SH 機能説明
(大体が
.B Vz
と同じなので独自とは言えませんが)
.B ne
独特のフィーチャーを説明します。
キー操作の説明に例示するキーは全て標準設定に従っています。
.\"
.sh 行編集とテキストの変更フラグ
行の中で編集を行ってもその時点ではまだ修正した事にはなりません。
カーソル移動のキー操作等によりその行を離れたり、
ファイルセーブ等のキー操作を実行した時点で始めて
その行が更新したのかを調べます。

つまり間違った文字入力や削除を行ってしまっても行移動するまでならば、
元の文字列に戻しさえすれば修正した事にはなりません。
この時、行編集の取り消しのキー操作 (\fB^QL\fP) により、
行編集前の状態に戻す事も可能です。
.\"
.sh キーボードマクロ
一連のキー操作をまとめて一つのキーに記憶させる機能です。
キーボードマクロ登録開始のキー操作 (\fB^_\fP) により登録を開始し、
覚えさせる一連のキー操作を行った後に、
キー操作として登録されていないキーを押した時点で、
そのキーにそのマクロを割り当てる事が可能です。
.\"
.sh カーソル位置保持
カーソルを上下に動かした時でカーソルが行末以降に来る場合、
フリーカーソルモードを有効にしているならば、
カーソルは行末を越えても同じ位置にあり、
あまりカーソルが左右に揺れて見にくい状態になる事を回避します。

フリーカーソルモードが無効の場合でも行末を越える場合以外でも、
なるべくカーソル位置が変わらないようになっています。
.\"
.sh ページングモード
ページ移動のキー操作 (\fB^R\fP、\fB^C\fP) にはページングモードが存在します。
システムガイド行の左端の一文字はこのモードを指し、
.B P
の場合は普通のページ切替えを意味し、
.B S
の場合は、ページ移動のキー操作で前後への検索を行います。

ページングモード切替えのキー操作 (\fB^@\fP) によって、
.B P
と
.B S
を切替える事が可能です。
.\"
.sh 文字アンドゥー
文字を削除するキー操作 (\fB[DEL]\fPなど) で削除した文字・文字列は、
削除文字バッファにスタックされていきます。
このバッファにスタックされた文字は
削除アンドゥーのキー操作 (\fB^U\fP) により復活させる事が可能です。
この場合、削除文字バッファが溢れない限りは全て復活できます。
ただし、上書きを行った文字はこの文字アンドゥーでは復活できません。
.\"
.sh ブロックバッファ
.B ne
はブロックバッファもスタック構造を持っています。
つまり、ブロックカットのキー操作 (\fB^Y\fP)、
ブロックヤンクのキー操作 (\fB^KK\fP) は、
指定した行範囲をブロックバッファにスタックします。

このスタックされたブロックバッファは、
積んだ順序と逆の順序でテキストに張り付ける事が可能です。
.\"
.sh バックアップファイル
バックアップ機能を有効にしている場合、ファイルを書き出す時に、
元のファイルのバックアップをとる機能です。
バックアップファイルには、元のファイル名 (拡張子含む) に、
"\fB.bak\fP"
を付けたファイルにリネームされます。
.\"
.sh プロファイル
任意の時点での編集状態 (編集しているテキストのファイル名・内容など) を、
編集対象のファイルとは別の、
.B プロファイル
という保存形式で保存する事が可能です。
プロファイル保存のキー操作 (\fB[ESC]W\fP) や、
終了のキー操作 (\fB[ESC]Q\fP)、
そしてシグナル等による強制終了時にプロファイルが保存されます。

.B ne -r
として起動するとこの保存された
.B プロファイル
 を読みだし、保存された時点の編集状態に戻す事が可能です。
.\"
.sh 漢字コード変換
ファイルの読みだし時は自動的にそのファイルの漢字コードを判別し、
内部表現である EUC に変換します。(いわゆる) 半角カナの変換も対応しています。

ファイルセーブ時は読み出した元の漢字コードにより書き出し、
編集中にもこの漢字コードの変更が可能です。
.\"
.sh ファイラ
ファイルオープンのキー操作 (\fB[ESC]O\fPなど) でファイルを開く場合、
ファイル名を入力するプロンプトで、
そのままリターンキーを押すとファイラが立ち上がります。
UNIX のフリーソフトウェアでは
.B FD clone
や
.B mc
が有名ですがこれに似たような物と考えてください。
実際のところは
.B Vz
に内蔵のファイラに近いのですがメニューなどの問題で、
実際には PC-98x1 のフリーソフトウェア
.B FL
の
.B FDモード
に一番近いかもしれません。
.\"
.sh 端末への対応の強化
端末の処理に
.B TERMCAP
を利用し、
エントリが足りない場合は VT200 相当のシーケンスを使用します。
それだけではどうしても対応できない環境の為に、
起動時に term.$NE_TERM と term.$TERM を読みだします。
現時点では作者の利用している
.B rxvt
用の設定を付属していますが、
これは シフト+カーソルキー や コントロール+カーソルキー が使用できます。
みなさんも有効な設定ができましたら是非とも御一報ください。
.\"
.\"
.\"
.SH	ユーザー・カスタマイズ
.B ne
が使用する設定ファイルでパスが指定されていない場合、
まずホームディレクトリの下の
.B .ne
ディレクトリ (\fB~/.ne/\fP) を調べます。
次にシステム標準のディレクトリ (\fB@prefix@/etc/ne/\fP) を調べます。
つまり、各ユーザーのホームの下の
.B .ne
ディレクトリが優先されます。
これを設定ファイルの検索パスと呼びます。

パスを含めた場合は、設定ファイルの検索パスは調べられません。

.B ne
は起動時に設定ファイルのパス検索の
.I nerc
を読みます。
つまり最初に
.B ~/.ne/nerc
を探し、無ければ
.B @prefix@/etc/ne/nerc
を探します。

.B ne
の標準の設定は
.B Vz
ライクな
.B /usr/local/etc/ne/nerc
です。
これを
.B ~/.ne/nerc
にコピーしてから設定の変更をしてみてください。
この設定ファイルではオプション設定のみを行い、
キー操作の設定や端末設定を行うファイルを呼び出すだけにしています。
もちろん
.B nerc
のみにこれら全ての設定をいれる事も可能ですがお勧めはしません。
使用する設定ファイル群を
.B ~/.ne
の下にコピーしてから、設定の変更を行い、
.B nerc
から呼び出すだけにする方がいいでしょう。

設定ファイルの基本的な記述方法はUNIXでよくある行指向の物です。
.B #
以降行末までをコメントとします。

制御命令は以下の通り:
.IP \fB!I\fP\ \fIfile\fP:
.I file
で示されたファイルを読み込みそれを設定として使います。
そのファイルを処理し終った後はこの次の行から再開します。
この時パスが指定されていない場合はまず ~/.ne/\fIfile\fP を読もうとし、
見つからなければ次に @prefix@/etc/ne/\fIfile\fP を読みにいきます。
.\"
.IP \fB!R\ V\fP:
次の行以降でオプション設定を行います。
\."
.IP \fB!R\ D\fP:
次の行以降で \fB[BS]\fP や \fB[DEL]\fP、\fB[UP]\fP 
などの特殊文字を定義します。
これにより
.B termcap
で吸収しきれない端末依存の問題に対処できます。
.\"
.IP \fB!R\ K[:\fIregion\fP\fB]\fP:
次の行以降でキーアサイン、つまりキー操作の設定を行います。
.I region
を省略した場合、エディタのメインでのキーアサインで、
.B sys
を指定した場合、メニューなどで使用されるキーアサインで、
.B eff
を指定した場合、ファイラーで使用されるキーアサインを意味します。
.br
.PP
変数は以下の通り:
.IP \fB$(\fIname\fP\fB)\fP:
内部変数 \fIname\fP に置換します。

.IP \fB$*(\fIname\fP\fB)\fP:
環境変数 \fIname\fP に置換します。

.IP \fB$?(\fIhex\fP\fB)\fP:
16進数の値 \fIhex\fP をそのままキャラクタコードとして置換します。

.\"
.\"
.SH オプション
オプション設定で設定できるオプションです。
[]の中に省略した場合のデフォルトを示します。
.\"
.ta 1i
.IP \fBAnsiColor\fP	[On]
ANSIエスケープシーケンスによる色を用います。
.IP \fBAutoIndent\fP	[On]
オートインデントを使用します。
.IP \fBBackfile\fP	[Off]
バックアップファイルを作成します。
.IP \fBCRMark\fP	[Off]
各行の終りのリターンコードを見えるように表示します。
ただし、AnsiColor が Off の場合は見づらくなるでしょう。
.IP \fBFreeCursor\fP	[On]
カーソルが行末尾以降にも移動するフリーカーソルモードを使用します。
.IP \fBJapanese\fP	[On]
メニューやメッセージを日本語で表示します。
.IP \fBNoCase\fP	[Off]
検索の際、大文字・小文字の区別を行いません。
.IP \fBNumber\fP	[Off]
行番号を表示します。
.IP \fBOverWrite\fP	[Off]
カーソル位置に上書きをします。
.IP \fBTabMark\fP	[Off]
タブコードを見えるように表示します。
.IP \fBTabStop\fP	[8]
タブカラムを設定します。
.\"
.\"
.SH キーアサイン
.\"
.sh ファイル処理:
.ta 1.2i
.\"
.IP \fB^KW\ [F02]	\fIfile_toggle\fP
現在編集中のファイルをトグル切替えます。
.\" 
.IP \fB^KT\ [F03]	\fIfile_select\fP
現在編集中のファイルをメニューで選択して切替えます。
.\" 
.IP \fB[ESC]O	\fIfile_open\fP
テキストをオープンします。
この時にファイルネームを聞いてくるのですが、
ここで何も入力せずにリターンを押すとファイラーモードに移ります。

既にオープンしたファイルを指定した場合は、
そのテキストが編集対象のテキストになります。
.\" 
.IP \fB[ESC]C	\fIfile_close\fP
対象のテキストをクローズします。
この時、修正フラグが立っている場合はセーブするかどうか聞いてきます。
\fBY\fP もしくは\fBリターン\fPを押せばセーブし、
\fBN\fP もしくは\fBスペース\fPでセーブしません。
この時 \fBESC\fP を押すとこの操作を取り消して、編集に戻ります。
.\" 
.IP \fB[ESC]L	\fIfile_copen\fP
現在編集中のファイルを閉じ、新しく別のファイルを開きます。
ただし、新しいファイルを開く事ができない場合、
元のファイルは閉じられません。
.\" 
.IP \fB[ESC]S	\fIfile_save\fP
現在編集中のファイルを保存します。
この時ファイルネームを変更する事が可能です。
.\" 
.IP \fB[ESC]Q	\fIfile_quit\fP
プロファイル情報をセーブした後そのまま終了します。
.\" 
.IP \fB[ESC]P	\fIfile_rename\fP
現在編集中のテキストのファイルネームを変更します。
変更を行うとファイルの変更フラグが立ちます。
.\" 
.IP \fB[ESC]W	\fIfile_profile\fP
プロファイルを保存します。
.\" 
.IP \fB[ESC]U	\fIfile_undo\fP
編集を放棄して同じファイル名のファイルを再度オープンします。
Vzと違いカーソル位置は移動しません。
.\" 
.IP \fB[ESC]I	\fIfile_insert\fP
カーソルのある行にファイルを挿入します。
.\" 
.IP \fB[ESC]X	\fIfile_aclose\fP
全てのファイルを一括クローズします。
変更フラグの立っているファイルのみ保存するかを確認をします。
.\"
.sh カーソル移動:
.\" 
.IP \fB^E\ [UP]	\fIcursor_up\fP
.IP \fB^X\ [DOWN]	\fIcursor_down\fP
.IP \fB^S\ [LEFT]	\fIcursor_left\fP
.IP \fB^D\ [RIGHT]	\fIcursor_right\fP
カーソルを上下左右へ移動します。
.\" 
.IP \fB^A	\fIcursor_tkprev\fP
.IP \fB^F	\fIcursor_tknext\fP
カーソルを前、次の単語の先頭へ移動します。
.\" 
.IP \fB^QS\ [HOME]	\fIcursor_sleft\fP
.IP \fB^QD\ [END]	\fIcursor_sright\fP
カーソルを行の先頭、最後へ移動します。
.\" 
.IP \fB^W	\fIcursor_rup\ [lines]\fP
.IP \fB^Z	\fIcursor_rdown\ [lines]\fP
表示しているテキストをそのまま上下にスクロールさせ、
カーソルの表示位置を変えずに行を移動させます。
.\" 
.IP \fB^R\ [PGUP]	\fIcursor_pup\fP
.IP \fB^C\ [PGDN]	\fIcursor_pdown\fP
ページングモードによってかわりますが、通常のページモードの場合は、
カーソルの表示位置を変えずにテキストを 1ページ上下に移動します。
この時 1ページとは行の数 -1 を意味します。
サーチモードの場合はそれぞれ上方向・下方向への検索を行います。
.\" 
.IP \fB^QE	\fIcursor_sup\fP
.IP \fB^QX	\fIcursor_sdown\fP
カーソルを画面の上端、下端へ移動します。
.\" 
.IP \fB^QR	\fIcursor_top\fP
.IP \fB^QC	\fIcursor_bottom\fP
カーソルをテキストの先頭、末尾に移動します。
行ジャンプ
.\" 
.IP \fB^QP	\fIjump_before\ [slot]\fP
直前に記憶したカーソル位置に移動します。
このカーソル位置の記憶は、^QR ^QC ^QF ^QM ^QP ^QJ ^Q1〜4で行われます。
.\" 
.IP \fB^QM	\fIjump_mark\ [slot]\fP
今のカーソル位置を記憶します。
.\" 
.IP \fB^QJ	\fIjump_line\ [line]\fP
指定した行番号にカーソルを移動します。
.\" 
.IP \fB^K1〜4	\fIjump_mark\ [slot]\fP
今のカーソル位置をそれぞれ1番〜4番のマークスロットに登録します。
.\" 
.IP \fB^Q1〜4	\fIjump_before\ [slot]\fP
それぞれ1番〜4番のマークスロットに登録された位置にカーソルを移動します。
.\" 
.IP \fB^QG	\fIjump_tag\fP
カーソルの指してる行にタグ情報が含まれている場合、
そのタグ情報の指すファイル・行番号に移動します。
.\"
.sh 編集:
.\" 
.IP \fB^M\ [CR]	\fIline_cr\fP
カーソルの位置でその行を切ってリターンコードを挿入し、
残りを次の新しい行にします。
アートインデントが有効ならば次の新しい行には、
元の行の先頭のスペース・タブが移されます。
.\" 
.IP \fB^U	\fIchar_undo\fP
直前に削除した文字や単語をカーソルの位置に挿入させます。
.\" 
.IP \fB^P	\fIchar_input\fP
制御文字を入力します。
.\" 
.IP \fB^N	\fIline_new\fP
カーソルの位置 (正確にはカーソルのある行の直前) に新しい行を挿入します。
.\" 
.IP \fB^QL	\fIline_undo\fP
行編集を取り消し、編集前の状態にします。
.\"
.sh 削除:
.\" 
.IP \fB^H\ [BS]	\fIdel_bs\fP
カーソルの直前の一文字を削除します。
.\" 
.IP \fB^G\ [DEL]	\fIdel_char\fP
カーソルの位置の一文字を削除します。
.\" 
.IP \fB^QH	\fIdel_tkprev\fP
.IP \fB^T	\fIdel_tknext\fP
それぞれカーソルの直前・位置の単語を削除します。
.\" 
.IP \fB^QT	\fIdel_sleft\fP
.IP \fB^QY	\fIdel_sright\fP
それぞれカーソルから行頭・行末までを削除します。
ブロック処理
.\" 
.IP \fB^B\ [F10]	\fIblock_start\fP
カーソル位置 (正確にはその行の直前) を基準とした
ブロック指定モードになります。
.\" 
.IP \fB^Y\ [F08]	\fIblock_cut\fP
カーソルの行もしくは、ブロック指定モードの場合そのブロックを、
編集テキストから削除し、ブロックバッファにスタックしていきます。
.\" 
.IP \fB^J\ [F09]	\fIblock_paste\fP
ブロックバッファにスタックされた文字列があるならば、
カーソルの位置に挿入し、そのスタックを破棄します。
.\" 
.IP \fB^KK	\fIblock_yank\fP
カーソルの行もしくは、ブロック指定モードの場合そのブロックを、
編集テキストを変更せずに、ブロックバッファにスタックしていきます。
.B ^Y
.I block_cut
と違うのは編集テキストを変更しない点です。
.\" 
.IP \fB^KC	\fIblock_copy\fP
ブロックバッファにスタックされた文字列があるならば、
カーソルの位置に挿入します。
.B ^J
.I block_paste
と違う点は、挿入したスタックはそのまま変更しない点です。
.\" 
.IP \fB^KY	\fIblock_kill\fP
ブロックバッファの全てのスタックを破棄します。
.\" 
.IP \fB^KY	\fIblock_chlast\fP
ブロックの先頭と末尾を交換し移動します。
.\" 
.IP \fB^KD	\fIblock_dup\fP
カーソルの行もしくは、ブロック指定モードの場合そのブロックを、
そのまま複写し編集テキストに挿入します。
.\"
.sh 検索・置換
.\" 
.IP \fB^@	\fIsearch_paging\fP
ページングモードを切替えます。
通常のページモードとサーチモードのみがあります。
ハーフページやタイトルサーチなどの機能はありません。
.\" 
.IP \fB^QF\ [F06]	\fIsearch_in\fP
検索する文字列を指定します。
この時文字列を指定するのみで検索はしません。
文字列が指定された場合はページングモードをサーチモードにします。
.\" 
.IP \fB^QA\ [F07]	\fIsearch_repl\fP
文字列を置換します。
.\" 
.IP \fB^QO	\fIsearch_repl_redo\fP
直前に行った文字列置換を再実行します。
.\" 
.IP \fB^L\ [F05]	\fIsearch_getword\fP
カーソル位置の単語を検索文字列として取得します。
.\" 
.IP		\fIsearch_prev\fP
.IP		\fIsearch_next\fP
上方向・下方向への検索を行います。
.\"
.sh 雑多な処理:
.\" 
.IP \fB^_	\fImisc_kmacro\fP
キーボードマクロの設定を開始します。
.\" 
.IP \fB^V\ [INS]	\fIopt_set\ "OverWrite"\fP
挿入モードと上書きモードを切替えます。
.\" 
.IP \fB[ESC]E	\fImisc_exec\fP
子プロセスを実行します。
.\" 
.IP \fB[F01]	\fImenu_file\fP
ファイル操作のメニューを出します。
.\" 
.IP \fB[F04]	\fImenu_option\fP
オプション設定・確認のメニューを出します。
.\" 
.IP \fB^KI	\fIopt_tab\fP
タブのカラム数を指定します。
.\"  toggle_indent  ^KH
.\"    オートインデントの有効・無効を切替えます。
.\" 
.IP \fB[ESC]D	\fImisc_redraw\fP
画面を際描画します。
.\"                    opt_linenum
.\"                    opt_kanji
.\"                    opt_retmode

.sh メニュー・文字列入力時:
.\" 
.IP \fB[ESC]	\fIescape\fP
メニューなどを終了します。
.\" 
.IP \fB[CR]	\fIreturn\fP
メニューの選択、文字列入力の決定などをします。
.\" 
.IP \fB^E\ [UP]	\fIcursor_up\fP
.IP \fB^X\ [DOWN]	\fIcursor_down\fP
メニューの場合はメニューカーソルを上下に移動します。
文字列入力時はヒストリーの選択を行います。
.\" 
.IP \fB^S\ [LEFT]	\fIcursor_left\fP
.IP \fB^D\ [RIGHT]	\fIcursor_right\fP
文字列入力時は、カーソルを左右に移動します。
.\" 
.IP \fB^R	\fIcursor_pup\fP
.IP \fB^C	\fIcursor_pdown\fP
メニューの場合、上下にページ移動します。
.\" 
.IP \fB^W	\fIcursor_rup\fP
.IP \fB^Z	\fIcursor_rdown\fP
メニューの場合、上下にロール移動します。
.\" 
.IP \fB^QR	\fIcursor_top\fP
.IP \fB^QC	\fIcursor_bottom\fP
メニューの場合、メニューの上端・下端に移動します。
.\" 
.IP \fB^QS\ [HOME]	\fIcursor_sleft\fP
.IP \fB^QD\ [END]	\fIcursor_sright\fP
文字列入力時、カーソルを左端・右端に移動します。

src/ne.c
/*--------------------------------------------------------------------
	nxeditor
			FILE NAME:nxedit.c
			Programed by : I.Neva
			R & D  ADVANCED SYSTEMS. IMAGING PRODUCTS.
			1992.06.01

    Copyright (c) 1998,1999,2000 SASAKI Shunsuke.
    All rights reserved. 
--------------------------------------------------------------------*/
#define	VAL_impl

#include "ed.h"
#include "sh.h"
#include <ctype.h>

#include	<errno.h>
#include	<sys/types.h>
#include	<sys/stat.h>
#include	<pwd.h>
#include	<fcntl.h>

#if HAVE_SYS_WAIT_H
#	include <sys/wait.h>
#endif

#ifdef	HAVE_SYS_UTSNAME_H
#	include	<sys/utsname.h>
#endif


	char *title1 = "        Multi FILE Editor      ";
	char *title2 = "               Ne              ";
	char *copyright = "Copyright (c) 1991, 1992 I.Nebashi";

static	void	opening()
{
/* Emacs 的ガイド表示をいれる。 */

//	term_cls();
//	term_color_normal();

//	term_flush();

//	term_puts("ne " VER);
//	term_inkey();
}


/*-------------------------------------------------------------------
	Main Command Loop
-------------------------------------------------------------------*/

#ifdef	__dead2
void	ne_loop(int region) __dead2;
#endif

void	ne_loop(int region)
{
 	int 	key;

	for (;;) 
		{
		 if (region==0)
		 	CrtDrawAll();

		 dsp_allview();
		 term_csrn();
		 key=get_keyf(region);
		 RefreshMessage();

		 if (region==0&& csrse.gf)
		 	{
		 	 csrse.gf=FALSE;
		 	 term_locate(GetRow(), GetCol()+NumWidth);
		 	 crt_crmark();
		 	}

		 if (key==-1)
		 	continue;

		 if ((key&KF_normalcode)==0)
		 	funclist[0][key](); else
		 	{
		 	 if (region==0)
		 	 	InputAndCrt(key& ~KF_normalcode); 
		 	}
		}
	/*NOTREACHED*/
}

void	ne_fin()
{
	exit(0);
	/*NOTREACHED*/
}


void	sysinfo_path(char *s,const char *t)
{
	sprintf(s, "%s/%s", sysinfo.nxpath, t);
}

void	ne_init()
{
	char	*p;

	sysinfo.vp_def=hash_init(NULL, MAX_val);
	opt_default();

	p=getenv("HOME");
	if (p!=NULL)
		strcpy(sysinfo.nxpath, p); else
		{
		 struct passwd	*pw;
		 pw=getpwuid(geteuid());
		 if (pw!=NULL)
		 	strcpy(sysinfo.nxpath, pw->pw_dir); else
			getcwd(sysinfo.nxpath, LN_path);
		}
	strcat(sysinfo.nxpath, "/.ne");
	mkdir(sysinfo.nxpath, 0777); // !!

	sysinfo.shell=getenv("SHELL");
	if (sysinfo.shell==NULL)
		sysinfo.shell="/bin/sh";

	config_read("nerc");
}

void	ne_report()
{
#ifndef	HAVE_UNAME
	report_puts("ne version " VER "\n");

#else
	struct utsname	u;

	uname(&u);
	report_printf("ne version " VER "  %s %s (%s)\n"
		, u.sysname, u.release, u.version);

#endif

	report_puts("***\n");
	report_puts("環境変数:\n");
	report_printf("  $HOME: %s\n", getenv("HOME"));
	report_printf("  $TERM: %s\n", getenv("TERM"));
	report_printf("  $NE_TERM: %s\n", getenv("NE_TERM"));
	report_printf("  $SHELL: %s\n", getenv("SHELL"));
	report_puts("\n");

	hash_report(sysinfo.vp_def);
	report_puts("\n");

	term_report();
	report_puts("\n");

	keydef_report();
	report_puts("***\n");

}

bool	ne_arg(int argc,char *argv[])
{
	int 	line;
	int 	optcount;
	bool	pf;
	char	buf[LN_dspbuf+1];
	bool	f;
	int 	c;
	char	*sp,*p;

	line=0;
	f=FALSE;
	pf=FALSE;
	sysinfo.f_report=FALSE;

	for (optcount=1; optcount<argc; ++optcount) // ????
		{
		 c=getopt(argc, argv, "Vrjeod:D:");
		 if (c==EOF)
		 	break;

		 switch(c) 
		 	{
		  case 'V':
		  	 sysinfo.f_report=TRUE;
		  	 report_set(stderr);
		  	 break;
		  case 'j':
		  	 hash_set(sysinfo.vp_def, "japanese","true");
		  	 break;
		  case 'e':
		  	 hash_set(sysinfo.vp_def, "japanese","false");
		  	 break;
		  case 'r':
		  	 pf=TRUE;
		  	 break;
		  case 'o':
		  	 opening();
		  	 break;
		  case 'D':
		  	 strcpy(buf,optarg);
		  	 optcount++;
		  	 sp=buf;
		  	 p=strsep(&sp,"=");
		  	 hash_set(sysinfo.vp_def, buf, p);
		 	}
		}

	sysinfo_optset();

	if (sysinfo.f_report)
		return FALSE;

	if (pf)
		return profile_read();

	for (; optcount<argc; ++optcount)
		{
		 if (*argv[optcount]=='+')
		 	line=atoi(argv[optcount]+1); else
		 	{
		 	 if (FileOpenOp(argv[optcount]))
		 	 	f=TRUE;
		 	}
		}

	if (!f&& line>0)
		{
		 int	a;

		 a=min(line,GetRowWidth()/2+1);

		 csr_setly(line-a+1);
		 csr_setdy(a);
		}

	if (!f)
		FileOpenOp("untitled.txt");

	return f;
}

int 	main(int argc,char *argv[])
{
	term_init();
	term_start();
	term_cls();
	dsp_allinit();
	SetSignal();

	lists_init();
	edbuf_init();

	keydef_init();
	ne_init();
	ne_arg(argc, argv);
	key_set();
	sysinfo_optset();

	if (sysinfo.f_report)
		{
		 ne_report();
		 exit(0);
		}

	udbuf_init();
	search_init();
	eff_init(NULL,NULL);
	system_guide_init();
	*sysinfo.doublekey='\0'; // !!

	ne_loop(0);
	/* NOTREACHED */
}



/*-------------------------------------------------------------------
	Escape Shell
-------------------------------------------------------------------*/
void	CommandCom(char *sys_buff)
{
	bool	f;

	term_stop();

	if (*sys_buff == '\0')
		{
		 puts(TYPE_EXIT_MSG);
		 system(sysinfo.shell);
		 f=FALSE;
		} else 
		{
		 puts(sys_buff);
		 system(sys_buff);
		 f=TRUE;
		}

	term_start();

	if (f)
		{
		 fputs(HIT_ANY_KEY_MSG, stdout);
		 fflush(stdout);
		 term_kflush();
		 term_inkey();
		}

	term_cls();
}

SHELL	void	op_misc_exec()
{
	char	buf[MAXEDITLINE+1];

	*buf='\0';
	if (HisGets(buf, GETS_SHELL_MSG, SHELLS_SYSTEM) != NULL)
		CommandCom(buf);
}


SHELL	void	op_opt_linenum()
{
	opt_set("number", NULL);

//	OffsetSetByColumn();
}

SHELL	void	op_misc_redraw()
{
	term_cls();
}



/*-----------------------------------------------------------------------------
	fork shell and take its stdout/stderr.
*/
SHELL	void	op_misc_insert_output( void )
{
	pid_t	pid_child ;
	int		pipefds[2] ;
	char	buf[MAXEDITLINE+1]="";


	if (HisGets(buf, GETS_SHELL_MSG, SHELLS_SYSTEM) == NULL)
		return ;
		/* NOTREACHED */

	/*	Normally, this may not be failed, but... */
	if (pipe(pipefds)!=0)
		{
		 system_msg(strerror(errno));
		 term_inkey();
		 return;
		/* NOTREACHED */
		}

	pid_child=fork();
	switch(pid_child)
		{
		/*	now we are child. */
		case	0 :
			/*	pipe plumbing.
				child process's stdout and stderr are dup'ed from same fd.
				whatever written into these will end up to our pipefds[ 0 ].
			*/
//			term_stop();

			close( 1 ) ;
			dup( pipefds[ 1 ] ) ;	/* duplicate to stdout */

			close( 2 ) ; 
			dup( pipefds[ 1 ] ) ;	/* duplicate to stderr */

			/*	these fds are no longer useful. */
			close( pipefds[ 0 ] ) ;
			close( pipefds[ 1 ] ) ;

			execl(sysinfo.shell, sysinfo.shell, "-c", buf, NULL) ;	/* should not be failed. */
			_exit( 1 ) ;
			/* NOTREACHED */

		/*	cannot be forked off. */
		case	-1 :
			system_msg(strerror(errno));
			term_inkey();
			system_msg("");

			close(pipefds[0]);
			close(pipefds[1]);

			break;
			/* NOTREACHED */

	/*	now we are parent. */
		default :
			{
			int	status ;
			FILE	*fp_pipe;

			close(pipefds[1]);
			fp_pipe = fdopen(pipefds[ 0 ], "r");
			if (fp_pipe==NULL)
				{
				 system_msg(strerror(errno));
				 term_inkey();
				 system_msg("");
				} else
				{
				 EditLine	*ed, *ed_new;
				 bool	f;
				 int 	n;

				 csr_leupdate();
				 ed_new=ed=GetList(GetLineOffset()-1);
				 f=FALSE;

				 term_stop();

				 while ( fgets(buf, sizeof(buf) - 1, fp_pipe) )
				 	{
				 	/* echo to stdout for user friendliness. :) */
				 	fputs(buf, stdout);

				 	/* remove last '\n' */
				 	n=strlen(buf)-1;
				 	if (buf[n] == '\n')
				 		buf[n] = '\0' ;

				 	/*	create new linebuffer and insert into list. */
				 	InsertLine(ed, MakeLine(buf));
				 	ed=ed->next;
				 	f=TRUE;
				 	}
				 fclose( fp_pipe ) ;
				 close(pipefds[0]);
				 term_start();

				 if (f)
				 	{
				 	 SetFileChangeFlag( ) ;	/* now, file is dirty. */
				 	 csrse.ed=ed_new->next;
				 	 csr_lenew();
				 	 OffsetSetByColumn();
				 	}
				}
			wait( &status ) ;

			term_cls();

			}
		}
}

